### PR TITLE
Tools: Generate bootloader hex using waf

### PR DESCRIPTION
**Pull request by sypaq-nexton**
The following workflow:
```
./waf configure --board <board> --bootloader
./waf bootloader
./waf configure --board <board>
./waf plane
```

will generate a new bootloader binary, but the new bootloader binary won't get copied to the `/Tools/bootloaders/` directory. As a result, the new bootloader binary _won't_ be combined with the plane binary. The required workflow is:
```
./Tools/scripts/build_bootloaders.py <board>
./waf configure --board <board>
./waf plane
```

This PR changes the `chibios.py` and `build_bootloaders.py` scripts so the two workflows are equivalent.

### CODE REVIEW

**Comment by sypaq-nexton** [comment]
@tridge, @peterbarker: FYI

File: Tools/ardupilotwaf/chibios.py [hunk 1]
```diff
@@ -211,14 +232,33 @@ def chibios_firmware(self):
...
+        copy_bin_task.set_run_after(generate_bin_task)
+
     if self.bld.env.HAVE_INTEL_HEX:
-        if os.path.exists(bootloader_bin.abspath()):
+        if self.env.BOOTLOADER:
             hex_target = self.bld.bldnode.find_or_declare('bin/' + link_output.change_ext('.hex').name)
-            hex_task = self.create_task('build_intel_hex', src=[bin_target, bootloader_bin], tgt=hex_target)
-            hex_task.set_run_after(generate_bin_task)
+            hex_task = self.create_task('build_bootloader_intel_hex', src=bin_target, tgt=hex_target)
+            hex_task.set_run_after(generate_apj_task)
```
Comment (khancyr) [inline Tools/ardupilotwaf/chibios.py:249]: why after the apj task and not bin task ?

Comment (sypaq-nexton) [reply Tools/ardupilotwaf/chibios.py]: You're right, shouldn't have been changed. Fixed in ec65aa92.

Comment (sypaq-nexton) [reply Tools/ardupilotwaf/chibios.py]: @khancyr: Any other changes you'd like me to make?

**Comment by tridge** [comment]
my issue with this is none of our waf build commands change any files which are checked into out git tree. That is quite deliberate, so we don't end up with a dirty tree when we do normal build commands.
We have Tools/scripts/build_bootloaders.py and Tools/scripts/build_iofirmware.py as separate commands outside of normal waf usage for convenience, but I'd prefer to not have it as part of standard waf usage.
sorry!

**Comment by sypaq-nexton** [comment]
@tridge: The core of the issue is that the vehicle build doesn't pull in a newly built bootloader unless it's published to the `/Tools/bootloaders/` directory. Maybe the solution is to update the vehicle build to look for bootloaders in the following order:
1. `/build/<board>/bin/`
2. `/Tools/bootloaders/`
3. `None`

rather than the current order:
1. `/Tools/bootloaders/`
2. `None`

**Comment by peterbarker** [comment]
On Thu, 18 Mar 2021, Nick Exton wrote:

> @tridge: The core of the issue is that the vehicle build doesn't pull in a newly built bootloader unless it's published to the /Tools/bootloaders/
> directory. Maybe the solution is to update the vehicle build to look for bootloaders in the following order:

That's actually quite intentional.  That extra step is there to try to 
prevent people from bricking their boards - it is quite difficult to 
recover some boards if the bootloader isn't working as things like boot0 
aren't exposed.

It's suggested that if you are doing work which impacts the bootloader 
then it is quite easy to formulate a compound command-line to do each of 
the steps in turn, so:

./waf configure --board=CubeBlack --bootloader && ./waf bootloader && cp 
build/CubeBlack/bin/AP_Bootloader.bin Tools/bootloaders/CubeBlack_bl.bin 
&& ./waf configure --board=CubeBlack && ./waf plane

**Comment by sypaq-nexton** [comment]
Thanks @peterbarker, that make sense. I forgot that hobbyists might not have the tools/expertise to recover a bricked board.

I'll close this out again.

**Pull request closed without merging**